After this fix you can toggle layers in tap dance without dalay.

diff --git a/quantum/quantum.c b/quantum/quantum.c
index 6530738b7..2b19c74c7 100644
--- a/quantum/quantum.c
+++ b/quantum/quantum.c
@@ -245,6 +245,7 @@ bool process_record_quantum(keyrecord_t *record) {
 
   #ifdef TAP_DANCE_ENABLE
     preprocess_tap_dance(keycode, record);
+    keycode = get_record_keycode(record);
   #endif
 
   if (!(
